h/static/scripts/annotator/anchoring/pdf.coffee

```diff
@@ -0,0 +1,224 @@
+      root = page.textLayer.textLayerDiv
+      selector = anchor.toSelector(options)
+      return html.anchor(root, [selector])
+    else
+      div = page.div ? page.el
+      placeholder = div.getElementsByClassName('annotator-placeholder')[0]
+      unless placeholder?
+        placeholder = document.createElement('span')
+        placeholder.classList.add('annotator-placeholder')
+        placeholder.textContent = 'Loading annotations…'
+        div.appendChild(placeholder)
+      range = document.createRange()
+      range.setStartBefore(placeholder)
+      range.setEndAfter(placeholder)
+      return range
+
+  if position?
+    promise = promise.catch ->
+      return findPage(position.start, cache.pageText)
+      .then ({index, offset, textContent}) ->
+        page = getPage(index)
+        start = position.start - offset
+        end = position.end - offset
+        length = end - start
+        assertQuote(textContent.substr(start, length))
+        anchor = new TextPositionAnchor(root, start, end)
+        return anchorByPosition(page, anchor)
+
+  if quote?
+    promise = promise.catch ->
+      {pagesCount} = PDFViewerApplication.pdfViewer
+
+      if position?
+        if cache.quotePosition[quote.exact]?[position.start]?
+         {page, anchor} = cache.quotePosition[quote.exact][position.start]
+         return anchorByPosition(page, anchor)
+
+      findInPages = ([pageIndex, rest...]) ->
+        page = getPage(pageIndex)
+        content = getPageTextContent(pageIndex, cache.pageText)
+        offset = getPageOffset(pageIndex, cache.pageText)
+        Promise.all([content, offset, page])
+        .then (results) ->
+          [content, offset, page] = results
+          root = {textContent: content}
+          pageOptions = {}
+          if position?
+            pageOptions.hint = position.start - offset
+          anchor = new TextQuoteAnchor.fromSelector(root, quote)
+          anchor = anchor.toPositionAnchor(pageOptions)
+          cache.quotePosition[quote.exact] ?= {}
+          cache.quotePosition[quote.exact][position.start] = {page, anchor}
+          return anchorByPosition(page, anchor)
+        .catch ->
+          if rest.length
+            return findInPages(rest)
+          else
+            throw new Error('quote not found')
+
+      pages = [0...pagesCount]
+
+      if position?
+        return findPage(position.start, cache.pageText)
+        .then ({index, offset, textContent}) ->
+          left = pages.slice(0, index)
+          right = pages.slice(index)
+          pages = []
+          while left.length or right.length
+            if right.length
+              pages.push(right.shift())
+            if left.length
+              pages.push(left.pop())
+          return findInPages(pages)
+      else
+        findInPages(pages)
+
+  return promise
+
+
+exports.describe = (root, range, options = {}) ->
+  cache = options.cache ? {}
+  range = new xpathRange.BrowserRange(range).normalize()
```

Comment:
Do we even need to use `xpathRange` here? Given that a native browser range has `startContainer` and `endContainer`, and each of these are guaranteed not to span a text layer boundary, won't that do?